### PR TITLE
[Concurrency] Disallow Sendable annotation on methods of non-Sendable types

### PR DESCRIPTION
**Pull request by angela-laar**
Disallow Sendable annotation on methods of non-Sendable types.

### CODE REVIEW

**Comment by angela-laar** [comment]
@swift-ci please smoke test

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (hborla) [inline include/swift/AST/DiagnosticsSema.def:1]: Nit: remove this newline

File: lib/Sema/TypeCheckAttr.cpp [hunk 2]
```diff
@@ -6625,6 +6627,16 @@ void AttributeChecker::visitSendableAttr(SendableAttr *attr) {
         .warnUntilSwiftVersion(6);
     }
   }
+  // Prevent Sendable Attr from being added to methods of non-sendable types
+  if (auto *funcDecl = dyn_cast<AbstractFunctionDecl>(D)) {
+    if (auto baseType = funcDecl->getImplicitSelfDecl()->getTypeInContext()) {
+      if (!isSendableType(dc->getParentModule(), baseType)) {
+        diagnoseAndRemoveAttr(
+                              attr, diag::nonsendable_instance_method)
```
Comment (hborla) [inline lib/Sema/TypeCheckAttr.cpp:6635]: Formatting nit: remove the line break between `(` and `attr`, e.g.

```
  diagnoseAndRemoveAttr(attr, diag::nonsendable_instance_method)
      .warnUntilSwiftVersion(6)
```

File: lib/Sema/TypeCheckAttr.cpp [hunk 3]
```diff
@@ -6625,6 +6627,16 @@ void AttributeChecker::visitSendableAttr(SendableAttr *attr) {
         .warnUntilSwiftVersion(6);
     }
   }
+  // Prevent Sendable Attr from being added to methods of non-sendable types
+  if (auto *funcDecl = dyn_cast<AbstractFunctionDecl>(D)) {
+    if (auto baseType = funcDecl->getImplicitSelfDecl()->getTypeInContext()) {
+      if (!isSendableType(dc->getParentModule(), baseType)) {
```
Comment (hborla) [inline lib/Sema/TypeCheckAttr.cpp:6633]: How does this behave for conditional conformances? e.g.

```swift
struct S<T> {
  let t: T

  @Sendable func test() {}
}

extension S: Sendable where T: Sendable {
  @Sendable func test2() {}
}
```

It seems like the first `@Sendable` should be diagnosed but the second should not.

Comment (slavapestov) [reply lib/Sema/TypeCheckAttr.cpp:6633]: It should work out. `isSendableType()` calls `TypeChecker::conformsToProtocol()` which checks conditional requirements, and `getSelfTypeInContext()` maps the interface type of `self` into the function's own generic environment, so the generic parameters become archetypes that satisfy those conditional requirements. Worth adding a test though!

File: lib/Sema/TypeCheckAttr.cpp [hunk 4]
```diff
@@ -6625,6 +6627,16 @@ void AttributeChecker::visitSendableAttr(SendableAttr *attr) {
         .warnUntilSwiftVersion(6);
     }
   }
+  // Prevent Sendable Attr from being added to methods of non-sendable types
+  if (auto *funcDecl = dyn_cast<AbstractFunctionDecl>(D)) {
+    if (auto baseType = funcDecl->getImplicitSelfDecl()->getTypeInContext()) {
```
Comment (slavapestov) [inline lib/Sema/TypeCheckAttr.cpp:6632]: This won't ever return null

**Comment by angela-laar** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckAttr.cpp (hunk 4 above)
Comment (angela-laar) [reply lib/Sema/TypeCheckAttr.cpp]: just noticed! updated!

**Comment by angela-laar** [comment]
@swift-ci please smoke test

**Comment by angela-laar** [comment]
@swift-ci please smoke test

**Review by hborla — APPROVED** [review]

**Comment by angela-laar** [comment]
@swift-ci please smoke test

**Comment by angela-laar** [comment]
@swift-ci please smoke test

**Pull request merged**